pecos/core/utils/matrix.hpp

```diff
@@ -607,6 +665,23 @@ namespace pecos {
     template<> inline float copy(ptrdiff_t *len, float *x, ptrdiff_t *xinc, float *y, ptrdiff_t *yinc) { return scopy_(len,x,xinc,y,yinc); }
 
     // ===== do_dot_product =====
+    template<class IX, class VX, class IY, class VY>
+    float32_t do_dot_product(const sparse_vec_t<IX, VX>& x, const sdvec_t<IY, VY>& y) {
+        float32_t ret = 0;
+        for(size_t s=0; s < x.nnz; s++) {
+            auto idx = x.idx[s];
```

Comment:
in other similar use cases (do_axpy), you are using `auto &`. Why such disparity?